Modified and implemented new backend services required for Contributor Admin Dashboard: GSoC M1.5

kshitij01042002 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #17440 .
2. This PR does the following: Introduced new backend services required for new Contributor Admin Dashboard. 
  Modified existing services in suggestion_services.py to update the new models introduced in #18340 . The initial load of these models will be populated using beam job introdued in #18406 .

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...",
followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] The linter/Karma presubmit checks have passed on my local machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [x] I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).


## Testing doc (for PRs with Beam jobs that modify production server data)

<!--
If this PR affects production server data, please follow
[these instructions](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production#submitting-a-pr-with-a-new-job-or-feature-that-requires-third-party-api)
and link to the job request doc here.

Otherwise, please delete this section.
-->

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- Make sure your PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

kshitij01042002 committed:
failing tests

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
restored changed file

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
added new services required for contributor admin dashboard

kshitij01042002 committed:
added services for translation updates

kshitij01042002 committed:
added services to update question stats

kshitij01042002 committed:
added tests for sevices in suggestion services

kshitij01042002 committed:
covered backend test coverage

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.5

kshitij01042002 committed:
added test to backend shrads file

kshitij01042002 committed:
Merge branch 'gsoc_M1.5' of https://github.com/kshitij01042002/oppia into gsoc_M1.5

kshitij01042002 commented:
Hi @seanlip @gp201 @chris7716 can you please take a pass at the PR? I need this to be merged as quickly as possible as my future PRs of GSoC depends on this and I might get delayed on the deadlines.
Thanks for coordinating.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.5

seanlip reviewed (approved):
LGTM for one codeowner file in core/storage. Sorry for missing that previously -- thanks for fixing it!

qinghaoyang reviewed (approved):
@kshitij01042002 LGTM.

JayVivarekar commented on core/domain/contributor_admin_dashboard_services_test.py:
```diff
@@ -0,0 +1,426 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests the Funtions to get stats Displayed in contributor Admin Dashboard."""
```
```suggestion
"""Tests the functions to get stats displayed in contributor Admin Dashboard."""
```

JayVivarekar commented on core/domain/contributor_admin_dashboard_services.py:
```diff
@@ -0,0 +1,458 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Funtions to get stats Displayed in contributor Admin Dashboard.
```
```suggestion
"""Functions to get stats displayed in contributor Admin Dashboard.
```

JayVivarekar reviewed (approved):
@kshitij01042002 Thanks! Minor nits otherwise LGTM

kshitij01042002 committed:
Update core/domain/contributor_admin_dashboard_services_test.py

Co-authored-by: Jay Vivarekar <jayviva@gmail.com>

kshitij01042002 committed:
Update core/domain/contributor_admin_dashboard_services.py

Co-authored-by: Jay Vivarekar <jayviva@gmail.com>

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.5

chris7716 commented on core/domain/suggestion_registry.py:
```diff
@@ -2003,28 +2003,6 @@ def __init__(
         self.submission_datetime = submission_datetime
 
 
-class TranslationSubmitterTotalContributionStatsDict(TypedDict):
```
Why are these being removed? Here and below.

kshitij01042002 replied:
Actually in PR #18358 on sagang's recommendation `to_dict` method was removed but I forget to delete the dict used. So this is basically dead code which isn't used anywhere.

chris7716 commented on core/domain/suggestion_services.py:
```diff
@@ -41,9 +42,8 @@
-    Sequence, Set, Tuple, Union, cast, overload
-)
+    Callable, Dict, Final, List, Literal, Mapping, Match,
+    Optional, Sequence, Set, Tuple, Union, cast, overload)
```
```suggestion
    Optional, Sequence, Set, Tuple, Union, cast, overload
)
```

kshitij01042002 replied:
pylint didn't allowed this, I made these changes only because I wasn't able to push code.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.5

chris7716 commented on core/domain/suggestion_services.py:
```diff
@@ -2484,10 +2484,171 @@ def _update_question_review_stats_models(
+            TranslationReviewerTotalContributionStats.
+            TranslationReviewerTotalContributionStats domain object.
+    """
+    stat = translation_reviewer_total_stat
```
Why would you use another variable? Can't you directly use `translation_submitter_total_stats`? Ditto elsewhere

kshitij01042002 replied:
Yeah we can directly use that, updated.

chris7716 commented on core/domain/suggestion_services.py:
```diff
@@ -2484,10 +2484,171 @@ def _update_question_review_stats_models(
+        translation_reviewer_total_stat.language_code,
+        translation_reviewer_total_stat.contributor_id)
+
+    assert stats_model is not None
```
Shouldn't we raise an exception if the model is not found? Ditto elsewhere.

kshitij01042002 replied:
I guess we don't need to do this because we are calling this method only when the model exists. If model doesn't exist we create a new model in `update_translation_review_stats`

chris7716 replied:
Okay. Then could you please explain it from a comment? Ditto elsewhere.

kshitij01042002 replied:
Updated with a comment. PTAL.

chris7716 commented on core/domain/suggestion_services.py:
```diff
@@ -2585,22 +2792,76 @@ def update_translation_contribution_stats_at_review(
+    rejected_translation_word_count = 0
+    accepted_translations_without_reviewer_edits_count = 0
+
+    if suggestion_is_accepted:
```
If `translation_contribution_stat_model` is not `None` we don't need to perform these checks since we already do it while incrementing stats.

kshitij01042002 replied:
Actually we need to do this is `translation_contribution_stat_model` is None or `translation_contribution_stat_model` is None so rather than doing it twice I got it out of the if condition.

chris7716 replied:
I hope you mean `translation_contribution_stat_model` is None or `translation_submitter_total_stat_model` is None cases

But still don't we repeat this here https://github.com/kshitij01042002/oppia/blob/9b206e29e401987d6d078a0f00fc4b44a0dbb4d2/core/domain/suggestion_services.py#L3385C5-L3385C5

Can we move to separate function for the case of `translation_contribution_stat_model` is None or `translation_submitter_total_stat_model` is None cases?

kshitij01042002 replied:
1. Yeah you are right, we need to calculate even if one of the case is None.

2. The code you have linked updates the existing model but if the model doesn't exist we have to calculate here.

3. I didn't quite get what exactly you mean?

chris7716 commented on core/domain/suggestion_services.py:
```diff
@@ -2808,11 +3170,69 @@ def update_question_contribution_stats_at_review(
+        accepted_questions_count = 0
+        accepted_questions_without_reviewer_edits_count = 0
+        rejected_questions_count = 0
+        if suggestion_is_accepted:
```
Don't need to perform these checks when `question_submitter_total_stat_model` is not `None`

kshitij01042002 replied:
Ditto above,

chris7716 commented on core/domain/suggestion_services.py:
```diff
@@ -2965,6 +3436,221 @@ def increment_question_review_stats(
+        if (
+                len(translation_submitter_total_stat
+                    .recent_review_outcomes)
+                > 99
```
Why 99? Can this be a constant somewhere?

kshitij01042002 replied:
Yeah that sounds good, Updated.

chris7716 reviewed (changes requested):
@kshitij01042002 Thanks! Took a pass.

kshitij01042002 committed:
changes requested by chris

kshitij01042002 committed:
Merge branch 'gsoc_M1.5' of https://github.com/kshitij01042002/oppia into gsoc_M1.5

kshitij01042002 committed:
removed reduntant variable

kshitij01042002 commented:
Hi @chris7716 thanks for the review, I have tried to address all your comments. PTAL.

gp201 reviewed (approved):
LGTM

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.5

aasiffaizal commented on core/domain/contributor_admin_dashboard_services.py:
```diff
@@ -0,0 +1,458 @@
+
+    translation_reviewer_stats = [
+        get_translation_reviewer_total_stats_from_model(model)
+        for model in translation_reviewer_models]
```
```suggestion
    translation_reviewer_stats = [
        get_translation_reviewer_total_stats_from_model(model)
        for model in translation_reviewer_models
    ]
```

aasiffaizal commented on core/domain/contributor_admin_dashboard_services.py:
```diff
@@ -0,0 +1,458 @@
+
+    question_submitter_stats = [
+        get_question_submitter_total_stats_from_model(model)
+        for model in question_submitter_models]
```
```suggestion
    question_submitter_stats = [
        get_question_submitter_total_stats_from_model(model)
        for model in question_submitter_models
    ]
```

aasiffaizal commented on core/domain/contributor_admin_dashboard_services.py:
```diff
@@ -0,0 +1,458 @@
+        )
+    )
+
+    question_reviewer_stats = [
+        get_question_reviewer_total_stats_from_model(model)
+        for model in question_reviewer_models]
```
```suggestion
    question_reviewer_stats = [
        get_question_reviewer_total_stats_from_model(model)
        for model in question_reviewer_models
    ]
```

aasiffaizal commented on core/domain/suggestion_services.py:
```diff
@@ -2518,6 +2698,51 @@ def update_translation_contribution_stats_at_submission(
+        translation_submitter_total_stat = (
+            contributor_admin_dashboard_services
+            .get_translation_submitter_total_stats_from_model(
+                translation_submitter_total_stat_model))
```
```suggestion
        translation_submitter_total_stat = (
            contributor_admin_dashboard_services
            .get_translation_submitter_total_stats_from_model(
                translation_submitter_total_stat_model
             )
         )
```

aasiffaizal commented on core/domain/suggestion_services.py:
```diff
@@ -2965,6 +3450,221 @@ def increment_question_review_stats(
+        if (
+                len(question_submitter_total_stat
+                    .recent_review_outcomes)
+                > 99
```
Can we have this as a constant? I see we have used the number 99 multiple times

kshitij01042002 replied:
Updated. PTAL.

aasiffaizal reviewed (changes requested):
Took a pass, left a few comments. Thanks @kshitij01042002

gp201 reviewed (approved):
LGTM for codeowner files

kshitij01042002 committed:
Update core/domain/contributor_admin_dashboard_services.py

Co-authored-by: Aasif Faizal <38973423+aasiffaizal@users.noreply.github.com>

kshitij01042002 committed:
Update core/domain/contributor_admin_dashboard_services.py

Co-authored-by: Aasif Faizal <38973423+aasiffaizal@users.noreply.github.com>

kshitij01042002 committed:
Update core/domain/contributor_admin_dashboard_services.py

Co-authored-by: Aasif Faizal <38973423+aasiffaizal@users.noreply.github.com>

kshitij01042002 committed:
Update core/domain/suggestion_services.py

Co-authored-by: Aasif Faizal <38973423+aasiffaizal@users.noreply.github.com>

kshitij01042002 committed:
made 99 as constant

kshitij01042002 committed:
Merge branch 'gsoc_M1.5' of https://github.com/kshitij01042002/oppia into gsoc_M1.5

kshitij01042002 committed:
changed name of fie

kshitij01042002 committed:
changed name in codeowner file

kshitij01042002 committed:
added test to backend shards

kshitij01042002 commented:
Hi @aasiffaizal addresssed your comments PTAL. Thanks for the review.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.5

aasiffaizal reviewed (approved):
LGTM

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.5

chris7716 reviewed (changes requested):
Thanks! Took another pass.

kshitij01042002 committed:
added comments for assertion

kshitij01042002 committed:
Merge branch 'gsoc_M1.5' of https://github.com/kshitij01042002/oppia into gsoc_M1.5

kshitij01042002 commented:
Hi @chris7716 thanks for the review. I have addressed your comments. Also I have some confusion so added it in the comment thread itself.PTAL. Thanks.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.5

kshitij01042002 committed:
moved code to a different function

kshitij01042002 committed:
fixed pylint errors

kshitij01042002 committed:
added docstring for the function

kshitij01042002 committed:
fixed pylint errors

kshitij01042002 committed:
fixed pylint errors

kshitij01042002 committed:
changed implementaion method for mypy checks

kshitij01042002 committed:
changed implementaion method for mypy checks

kshitij01042002 committed:
Merge branch 'gsoc_M1.5' of https://github.com/kshitij01042002/oppia into gsoc_M1.5

chris7716 reviewed (approved):
Thanks. LGTM!

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.5
